src/bp/nlu-core/training-worker-queue.ts

```diff
@@ -0,0 +1,253 @@
+import { NLU } from 'botpress/sdk'
+import cluster, { Worker, worker } from 'cluster'
+
+import { registerMsgHandler, spawnNewTrainingWorker, WORKER_TYPES } from '../cluster'
+
+import { initializeTools } from './initialize-tools'
+import { Trainer, TrainInput, TrainOutput } from './training-pipeline'
+
+type OutgoingPayload = Partial<{
+  config: NLU.Config
+  input: TrainInput
+}>
+type OutgoingMessageType = 'make_new_worker' | 'start_training' | 'cancel_training'
+interface OutgoingMessage {
+  type: OutgoingMessageType
+  payload: OutgoingPayload
+  destWid?: number
+}
+
+type Log = Partial<{ info: string; warning: string; error: string }>
+type IncomingPayload = Partial<{
+  log: Log
+  workerId: number
+  output: TrainOutput
+  error: string
+  progress: number
+}>
+type IncomingMessageType =
+  | 'log'
+  | 'worker_ready'
+  | 'training_canceled'
+  | 'training_done'
+  | 'training_progress'
+  | 'training_error'
+interface IncomingMessage {
+  type: IncomingMessageType
+  payload: IncomingPayload
+  srcWid: number
+}
+
+export class TrainingCanceledError extends Error {}
+
+export class TrainingWorkerQueue {
+  private waitingWorkers: number[] = []
+  private activeWorkers: { [trainSessionId: string]: number } = {}
```

Comment:
just so you know we have a Dic<T> shortcut that does the same. Knowing you, you probably made this on purpose though, explicitly specifying that the key is a trainSessionId